src/traces/streamtube/convert.js

```diff
@@ -0,0 +1,217 @@
+        var uvx = selection.data.velocity;
+
+        selection.traceCoordinate = [
+            fromDataScale(pos[0], 'xaxis'),
+            fromDataScale(pos[1], 'yaxis'),
+            fromDataScale(pos[2], 'zaxis'),
+
+            fromDataScale(uvx[0], 'xaxis'),
+            fromDataScale(uvx[1], 'yaxis'),
+            fromDataScale(uvx[2], 'zaxis'),
+
+            // u/v/w norm
+            selection.data.intensity * this.data._normMax,
+            // divergence
+            selection.data.divergence
+        ];
+
+        selection.textLabel = this.data.text;
+
+        return true;
+    }
+};
+
+function distinctVals(col) {
+    return Lib.distinctVals(col).vals;
+}
+
+function getBoundPads(vec) {
+    var len = vec.length;
+    if(len === 1) {
+        return [0.5, 0.5];
+    } else {
+        return [vec[1] - vec[0], vec[len - 1] - vec[len - 2]];
+    }
+}
+
+function convert(scene, trace) {
+    var sceneLayout = scene.fullSceneLayout;
+    var dataScale = scene.dataScale;
+    var len = trace._len;
+    var tubeOpts = {};
+
+    function toDataCoords(arr, axisName) {
+        var ax = sceneLayout[axisName];
+        var scale = dataScale[axisName2scaleIndex[axisName]];
+        return Lib.simpleMap(arr, function(v) { return ax.d2l(v) * scale; });
+    }
+
+    var u = toDataCoords(trace.u.slice(0, len), 'xaxis');
+    var v = toDataCoords(trace.v.slice(0, len), 'yaxis');
+    var w = toDataCoords(trace.w.slice(0, len), 'zaxis');
+
+    tubeOpts.vectors = zip3(u, v, w);
+
+    var valsx = distinctVals(trace.x.slice(0, len));
+    var valsy = distinctVals(trace.y.slice(0, len));
+    var valsz = distinctVals(trace.z.slice(0, len));
+
+    // Over-specified mesh case, this would error in tube2mesh
+    if(valsx.length * valsy.length * valsz.length > len) {
+        return {positions: [], cells: []};
+    }
+
+    var meshx = toDataCoords(valsx, 'xaxis');
+    var meshy = toDataCoords(valsy, 'yaxis');
+    var meshz = toDataCoords(valsz, 'zaxis');
+
+    tubeOpts.meshgrid = [meshx, meshy, meshz];
+
+    if(trace.startx && trace.starty && trace.startz) {
+        var slen = trace._slen;
+        tubeOpts.startingPositions = zip3(
+            toDataCoords(trace.startx.slice(0, slen), 'xaxis'),
+            toDataCoords(trace.starty.slice(0, slen), 'yaxis'),
+            toDataCoords(trace.startz.slice(0, slen), 'zaxis')
+        );
+    } else {
+        // Default starting positions: cut xz plane at min-y,
+        // takes all x/y/z pts on that plane except those on the edges
+        // to generate "well-defined" tubes
+        var sy0 = sceneLayout.yaxis.d2l(trace._ybnds[0]) * dataScale[1];
+        var sx, sz;
+        if(valsx.length > 2 && valsz.length > 2) {
+            sx = valsx.slice(1, valsx.length - 1);
+            sz = valsz.slice(1, valsy.length - 1);
+        } else {
+            // use all pts on 2x2 and 1x1 planes
+            sx = valsx.slice();
+            sz = valsz.slice();
```

Comment:
Presumably this `length > 2` condition should be done separately for x and z? Also, I don't know what model is used for extending the field outside the x/y/z bounds, but if it's uncertain enough that at 3+ points it's best not to start at the last coordinate, then I'd think at 2 points it'd be best to use a single point halfway between the two.

Reply:
> Also, I don't know what model is used for extending the field outside the x/y/z bounds

Currently, positions outside the mesh get (u=0, v=0, w=0) if I understand correctly:

https://github.com/gl-vis/gl-streamtube3d/blob/0359ed9696de5af570fa6a5e36bdfbca2013220b/streamtube.js#L228-L231

cc @kig

Reply:
New attempt in https://github.com/plotly/plotly.js/pull/2658/commits/83f3bdc71c3a3b20715f536f8f368e050656fa40